client/session.go

```diff
@@ -563,3 +595,945 @@ func (s *session) Close() error {
+		blocksMetadataQueues    []blocksMetadataQueue
+	)
+	for perPeerBlocksMetadata := range enqueueCh.get() {
+		// Filter and select which blocks to retrieve from which peers
+		if err := s.selectBlocksForSeriesFromPeerBlocksMetadata(
+			perPeerBlocksMetadata, peerQueues,
+			currStart, currEligible, blocksMetadataQueues,
+		); err != nil {
+			enqueueCh.trackProcessed(1)
+			continue
+		}
+
+		// Insert work into peer queues
+		var wg sync.WaitGroup
+		for _, peerBlocksMetadata := range perPeerBlocksMetadata {
+			if len(peerBlocksMetadata.blocks) == 0 {
+				// No blocks assigned
+				continue
+			}
+			queue := peerQueues.findQueue(peerBlocksMetadata.peer)
+			wg.Add(1)
+			queue.enqueue(peerBlocksMetadata, wg.Done)
+		}
+		// Launch asynchronous waiter to mark completion
+		go func() {
+			wg.Wait()
+			enqueueCh.trackProcessed(1)
+		}()
+	}
+
+	// Close all queues
+	peerQueues.closeAll()
+
+	return nil
+}
+
+func (s *session) streamCollectedBlocksMetadata(
+	peersLen int,
+	ch <-chan blocksMetadata,
+	enqueueCh *enqueueChannel,
+) {
+	metadata := make(map[string]*receivedBlocks)
+
+	// Receive off of metadata channel
+	for {
+		m, ok := <-ch
+		if !ok {
+			break
+		}
+
+		received, ok := metadata[m.id]
+		if !ok {
+			received = &receivedBlocks{
+				results: make([]*blocksMetadata, 0, peersLen),
+			}
+			metadata[m.id] = received
+		}
+		if received.submitted {
+			// Already submitted to enqueue channel
+			continue
+		}
+
+		received.results = append(received.results, &m)
+
+		if len(received.results) == peersLen {
+			enqueueCh.enqueue(received.results)
+			received.submitted = true
+		}
+	}
+
+	// Enqueue all unsubmitted received metadata
+	for _, received := range metadata {
+		if received.submitted {
+			continue
+		}
+		enqueueCh.enqueue(received.results)
+	}
+}
+
+func (s *session) selectBlocksForSeriesFromPeerBlocksMetadata(
+	perPeerBlocksMetadata []*blocksMetadata,
+	peerQueues peerBlocksQueues,
+	pooledCurrStart, pooledCurrEligible []*blocksMetadata,
+	pooledBlocksMetadataQueues []blocksMetadataQueue,
+) error {
+	// Use pooled arrays
+	var (
+		currStart            = pooledCurrStart[:0]
+		currEligible         = pooledCurrEligible[:0]
+		blocksMetadataQueues = pooledBlocksMetadataQueues[:0]
+	)
+
+	// Sort the per peer metadatas by peer ID for consistent results
+	sort.Sort(peerBlocksMetadata(perPeerBlocksMetadata))
+
+	// Sort the metadatas per peer by time and reset the selection index
+	for _, blocksMetadata := range perPeerBlocksMetadata {
+		sort.Sort(blockMetadatasByTime(blocksMetadata.blocks))
+		// Reset the selection index
+		blocksMetadata.idx = 0
+	}
+
+	// Select blocks from peers
+	for {
+		// Find the earliest start time
+		var earliestStart time.Time
+		for _, blocksMetadata := range perPeerBlocksMetadata {
+			if len(blocksMetadata.unselectedBlocks()) == 0 {
+				// No unselected blocks
+				continue
+			}
+			unselected := blocksMetadata.unselectedBlocks()
+			if earliestStart.Equal(time.Time{}) ||
```

Comment:
Take it or leave it, but this can be merged with the next block so it's more terse and we don't have to iterate over the metadata twice:
```
currStart = currStart[:0]
for _, blocksMetadata := range perPeerBlocksMetadata {
  if len(blocksMetadata.unselectedBlocks()) == 0 {
    // No unselected blocks
    continue
  }
  unselected := blocksMetadata.unselectedBlocks()
  if earliestStart.IsZero() || unselected[0].start.Before(earliestStart) {
    earliestStart = unselected[0].start
    currStart = currStart[:0]
    currStart = append(currStart, blocksMetadata)
  } else if unselected[0].start == earliestStart {
    currStart = append(currStart, blocksMetadata)
  }
}
```